File: tests/test_data/cdk_scenarios/typescript_constructs/lib/costpilot-test-stack.ts
import * as cdk from 'aws-cdk-lib';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as elbv2 from 'aws-cdk-lib/aws-elasticloadbalancingv2';
import * as autoscaling from 'aws-cdk-lib/aws-autoscaling';
import * as rds from 'aws-cdk-lib/aws-rds';
import * as elasticache from 'aws-cdk-lib/aws-elasticache';
import * as cloudwatch from 'aws-cdk-lib/aws-cloudwatch';
import { Construct } from 'constructs';

export interface CostPilotTestStackProps extends cdk.StackProps {
  environment: string;
}

export class CostPilotTestStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: CostPilotTestStackProps) {
    super(scope, id, props);

    const { environment } = props;

    // VPC with multiple subnets
    const vpc = new ec2.Vpc(this, 'TestVpc', {
      cidr: '10.0.0.0/16',
      maxAzs: 2,
      subnetConfiguration: [
        {
          cidrMask: 24,
          name: 'Public',
          subnetType: ec2.SubnetType.PUBLIC,
        },
        {
          cidrMask: 24,
          name: 'Private',
          subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS,
        },
      ],
      tags: {
        Name: `${environment}-VPC`,
        Environment: environment,
      },
    });

    // Security Groups
    const webSecurityGroup = new ec2.SecurityGroup(this, 'WebSecurityGroup', {
      vpc,
      description: 'Security group for web servers',
      allowAllOutbound: true,
    });

    webSecurityGroup.addIngressRule(
      ec2.Peer.anyIpv4(),
      ec2.Port.tcp(80),
      'Allow HTTP traffic'
    );

    webSecurityGroup.addIngressRule(
      ec2.Peer.anyIpv4(),
      ec2.Port.tcp(443),
      'Allow HTTPS traffic'
    );

    const dbSecurityGroup = new ec2.SecurityGroup(this, 'DatabaseSecurityGroup', {
      vpc,
      description: 'Security group for database',
      allowAllOutbound: true,
    });

    dbSecurityGroup.addIngressRule(
      webSecurityGroup,
      ec2.Port.tcp(3306),
      'Allow MySQL traffic from web servers'
    );

    // Application Load Balancer
    const alb = new elbv2.ApplicationLoadBalancer(this, 'ApplicationLoadBalancer', {
      vpc,
      internetFacing: true,
      securityGroup: new ec2.SecurityGroup(this, 'ALBSecurityGroup', {
        vpc,
        description: 'Security group for ALB',
        allowAllOutbound: true,
      }),
      tags: {
        Name: `${environment}-ALB`,
        Environment: environment,
      },
    });

    // ALB Security Group rules
    alb.connections.allowFromAnyIpv4(ec2.Port.tcp(80), 'Allow HTTP traffic');

    // Target Group
    const targetGroup = new elbv2.ApplicationTargetGroup(this, 'TargetGroup', {
      vpc,
      port: 80,
      protocol: elbv2.ApplicationProtocol.HTTP,
      targetType: elbv2.TargetType.INSTANCE,
      healthCheck: {
        path: '/',
        interval: cdk.Duration.seconds(30),
        timeout: cdk.Duration.seconds(5),
        healthyThresholdCount: 2,
        unhealthyThresholdCount: 2,
      },
    });

    // ALB Listener
    const listener = alb.addListener('HTTPListener', {
      port: 80,
      protocol: elbv2.ApplicationProtocol.HTTP,
      defaultAction: elbv2.ListenerAction.forward([targetGroup]),
    });

    // Launch Template
    const launchTemplate = new ec2.LaunchTemplate(this, 'LaunchTemplate', {
      instanceType: ec2.InstanceType.of(ec2.InstanceClass.T3, ec2.InstanceSize.MICRO),
      machineImage: ec2.MachineImage.latestAmazonLinux2023(),
      securityGroup: webSecurityGroup,
      userData: ec2.UserData.custom(`#!/bin/bash
        yum update -y
        yum install -y httpd
        systemctl start httpd
        systemctl enable httpd
        echo "<h1>Hello from ${environment} environment</h1>" > /var/www/html/index.html
      `),
      tagSpecifications: [
        {
          resourceType: ec2.LaunchTemplateResourceType.INSTANCE,
          tags: {
            Name: `${environment}-Web-Instance`,
            Environment: environment,
          },
        },
      ],
    });

    // Auto Scaling Group
    const autoScalingGroup = new autoscaling.AutoScalingGroup(this, 'AutoScalingGroup', {
      vpc,
      vpcSubnets: { subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS },
      launchTemplate,
      minCapacity: 2,
      maxCapacity: 10,
      desiredCapacity: 3,
      healthCheck: autoscaling.HealthCheck.ec2({
        grace: cdk.Duration.seconds(300),
      }),
      groupMetrics: [new autoscaling.GroupMetric(new cloudwatch.Metric({
        namespace: 'AWS/EC2',
        metricName: 'CPUUtilization',
        dimensionsMap: {
          AutoScalingGroupName: autoScalingGroup.autoScalingGroupName,
        },
      }))],
    });

    // Attach ASG to Target Group
    autoScalingGroup.attachToApplicationTargetGroup(targetGroup);

    // RDS Instance
    const rdsInstance = new rds.DatabaseInstance(this, 'DatabaseInstance', {
      vpc,
      vpcSubnets: { subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS },
      engine: rds.DatabaseInstanceEngine.mysql({
        version: rds.MysqlEngineVersion.VER_8_0,
      }),
      instanceType: ec2.InstanceType.of(ec2.InstanceClass.DB, ec2.InstanceSize.T3_MICRO),
      credentials: rds.Credentials.fromGeneratedSecret('admin'),
      securityGroups: [dbSecurityGroup],
      allocatedStorage: 20,
      storageType: rds.StorageType.GP2,
      backupRetention: cdk.Duration.days(7),
      multiAz: false,
      storageEncrypted: true,
      enablePerformanceInsights: true,
      performanceInsightRetention: rds.PerformanceInsightRetentionPeriod.DEFAULT,
      tags: {
        Name: `${environment}-RDS-Instance`,
        Environment: environment,
      },
    });

    // ElastiCache Cluster
    const cacheSubnetGroup = new elasticache.CfnSubnetGroup(this, 'CacheSubnetGroup', {
      description: `Subnet group for ${environment} ElastiCache`,
      subnetIds: vpc.selectSubnets({ subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS }).subnetIds,
      tags: [{
        key: 'Name',
        value: `${environment}-Cache-Subnet-Group`,
      }],
    });

    const cacheSecurityGroup = new ec2.SecurityGroup(this, 'CacheSecurityGroup', {
      vpc,
      description: 'Security group for ElastiCache',
      allowAllOutbound: true,
    });

    cacheSecurityGroup.addIngressRule(
      webSecurityGroup,
      ec2.Port.tcp(6379),
      'Allow Redis traffic from web servers'
    );

    const cacheCluster = new elasticache.CfnCacheCluster(this, 'CacheCluster', {
      cacheNodeType: 'cache.t3.micro',
      engine: 'redis',
      numCacheNodes: 1,
      cacheSubnetGroupName: cacheSubnetGroup.ref,
      vpcSecurityGroupIds: [cacheSecurityGroup.securityGroupId],
      tags: [{
        key: 'Name',
        value: `${environment}-ElastiCache-Cluster`,
      }],
    });

    // CloudWatch Alarms
    new cloudwatch.Alarm(this, 'CPUAlarm', {
      alarmName: `${environment}-cpu-alarm`,
      alarmDescription: 'Alarm for high CPU utilization',
      metric: new cloudwatch.Metric({
        namespace: 'AWS/EC2',
        metricName: 'CPUUtilization',
        dimensionsMap: {
          AutoScalingGroupName: autoScalingGroup.autoScalingGroupName,
        },
        statistic: 'Average',
      }),
      threshold: 80,
      evaluationPeriods: 2,
      comparisonOperator: cloudwatch.ComparisonOperator.GREATER_THAN_THRESHOLD,
    });

    new cloudwatch.Alarm(this, 'DatabaseCPUAlarm', {
      alarmName: `${environment}-db-cpu-alarm`,
      alarmDescription: 'Alarm for high database CPU utilization',
      metric: new cloudwatch.Metric({
        namespace: 'AWS/RDS',
        metricName: 'CPUUtilization',
        dimensionsMap: {
          DBInstanceIdentifier: rdsInstance.instanceIdentifier,
        },
        statistic: 'Average',
      }),
      threshold: 80,
      evaluationPeriods: 2,
      comparisonOperator: cloudwatch.ComparisonOperator.GREATER_THAN_THRESHOLD,
    });

    // Scaling Policy
    autoScalingGroup.scaleOnCpuUtilization('CpuScalingPolicy', {
      targetUtilizationPercent: 70,
    });

    // Outputs
    new cdk.CfnOutput(this, 'VpcId', {
      description: 'VPC ID',
      value: vpc.vpcId,
      exportName: `${environment}-VpcId`,
    });

    new cdk.CfnOutput(this, 'LoadBalancerDNS', {
      description: 'Load Balancer DNS Name',
      value: alb.loadBalancerDnsName,
      exportName: `${environment}-LoadBalancerDNS`,
    });

    new cdk.CfnOutput(this, 'DatabaseEndpoint', {
      description: 'Database Endpoint',
      value: rdsInstance.dbInstanceEndpointAddress,
      exportName: `${environment}-DatabaseEndpoint`,
    });

    new cdk.CfnOutput(this, 'CacheEndpoint', {
      description: 'Cache Cluster Endpoint',
      value: cacheCluster.attrRedisEndpointAddress,
      exportName: `${environment}-CacheEndpoint`,
    });
  }
}

File: tests/test_data/cdk_scenarios/python_constructs/costpilot_test_stack.py
from aws_cdk import (
    Duration,
    Stack,
    CfnOutput,
    aws_ec2 as ec2,
    aws_elasticloadbalancingv2 as elbv2,
    aws_autoscaling as autoscaling,
    aws_rds as rds,
    aws_elasticache as elasticache,
    aws_cloudwatch as cloudwatch,
)
from constructs import Construct


class CostPilotTestStack(Stack):

    def __init__(self, scope: Construct, construct_id: str, environment: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        self.environment = environment

        # VPC with multiple subnets
        vpc = ec2.Vpc(self, "TestVpc",
            cidr="10.0.0.0/16",
            max_azs=2,
            subnet_configuration=[
                ec2.SubnetConfiguration(
                    cidr_mask=24,
                    name="Public",
                    subnet_type=ec2.SubnetType.PUBLIC,
                ),
                ec2.SubnetConfiguration(
                    cidr_mask=24,
                    name="Private",
                    subnet_type=ec2.SubnetType.PRIVATE_WITH_EGRESS,
                ),
            ],
            tags={
                "Name": f"{environment}-VPC",
                "Environment": environment,
            }
        )

        # Security Groups
        web_security_group = ec2.SecurityGroup(self, "WebSecurityGroup",
            vpc=vpc,
            description="Security group for web servers",
            allow_all_outbound=True,
        )

        web_security_group.add_ingress_rule(
            ec2.Peer.any_ipv4(),
            ec2.Port.tcp(80),
            "Allow HTTP traffic"
        )

        web_security_group.add_ingress_rule(
            ec2.Peer.any_ipv4(),
            ec2.Port.tcp(443),
            "Allow HTTPS traffic"
        )

        db_security_group = ec2.SecurityGroup(self, "DatabaseSecurityGroup",
            vpc=vpc,
            description="Security group for database",
            allow_all_outbound=True,
        )

        db_security_group.add_ingress_rule(
            web_security_group,
            ec2.Port.tcp(3306),
            "Allow MySQL traffic from web servers"
        )

        # Application Load Balancer
        alb_security_group = ec2.SecurityGroup(self, "ALBSecurityGroup",
            vpc=vpc,
            description="Security group for ALB",
            allow_all_outbound=True,
        )

        alb = elbv2.ApplicationLoadBalancer(self, "ApplicationLoadBalancer",
            vpc=vpc,
            internet_facing=True,
            security_group=alb_security_group,
            tags={
                "Name": f"{environment}-ALB",
                "Environment": environment,
            }
        )

        # ALB Security Group rules
        alb.connections.allow_from_any_ipv4(ec2.Port.tcp(80), "Allow HTTP traffic")

        # Target Group
        target_group = elbv2.ApplicationTargetGroup(self, "TargetGroup",
            vpc=vpc,
            port=80,
            protocol=elbv2.ApplicationProtocol.HTTP,
            target_type=elbv2.TargetType.INSTANCE,
            health_check=elbv2.HealthCheck(
                path="/",
                interval=Duration.seconds(30),
                timeout=Duration.seconds(5),
                healthy_threshold_count=2,
                unhealthy_threshold_count=2,
            ),
        )

        # ALB Listener
        listener = alb.add_listener("HTTPListener",
            port=80,
            protocol=elbv2.ApplicationProtocol.HTTP,
            default_action=elbv2.ListenerAction.forward([target_group]),
        )

        # Launch Template
        launch_template = ec2.LaunchTemplate(self, "LaunchTemplate",
            instance_type=ec2.InstanceType.of(ec2.InstanceClass.T3, ec2.InstanceSize.MICRO),
            machine_image=ec2.MachineImage.latest_amazon_linux2023(),
            security_group=web_security_group,
            user_data=ec2.UserData.custom(f"""#!/bin/bash
                yum update -y
                yum install -y httpd
                systemctl start httpd
                systemctl enable httpd
                echo "<h1>Hello from {environment} environment</h1>" > /var/www/html/index.html
            """),
            tag_specifications=[
                ec2.LaunchTemplateTagSpecification(
                    resource_type=ec2.LaunchTemplateResourceType.INSTANCE,
                    tags={
                        "Name": f"{environment}-Web-Instance",
                        "Environment": environment,
                    }
                )
            ]
        )

        # Auto Scaling Group
        auto_scaling_group = autoscaling.AutoScalingGroup(self, "AutoScalingGroup",
            vpc=vpc,
            vpc_subnets=ec2.SubnetSelection(subnet_type=ec2.SubnetType.PRIVATE_WITH_EGRESS),
            launch_template=launch_template,
            min_capacity=2,
            max_capacity=10,
            desired_capacity=3,
            health_check=autoscaling.HealthCheck.ec2(
                grace=Duration.seconds(300),
            ),
        )

        # Attach ASG to Target Group
        auto_scaling_group.attach_to_application_target_group(target_group)

        # RDS Instance
        rds_instance = rds.DatabaseInstance(self, "DatabaseInstance",
            vpc=vpc,
            vpc_subnets=ec2.SubnetSelection(subnet_type=ec2.SubnetType.PRIVATE_WITH_EGRESS),
            engine=rds.DatabaseInstanceEngine.mysql(
                version=rds.MysqlEngineVersion.VER_8_0
            ),
            instance_type=ec2.InstanceType.of(ec2.InstanceClass.DB, ec2.InstanceSize.T3_MICRO),
            credentials=rds.Credentials.from_generated_secret("admin"),
            security_groups=[db_security_group],
            allocated_storage=20,
            storage_type=rds.StorageType.GP2,
            backup_retention=Duration.days(7),
            multi_az=False,
            storage_encrypted=True,
            enable_performance_insights=True,
            tags={
                "Name": f"{environment}-RDS-Instance",
                "Environment": environment,
            }
        )

        # ElastiCache Cluster
        cache_subnet_group = elasticache.CfnSubnetGroup(self, "CacheSubnetGroup",
            description=f"Subnet group for {environment} ElastiCache",
            subnet_ids=vpc.select_subnets(subnet_type=ec2.SubnetType.PRIVATE_WITH_EGRESS).subnet_ids,
            tags=[{
                "key": "Name",
                "value": f"{environment}-Cache-Subnet-Group",
            }]
        )

        cache_security_group = ec2.SecurityGroup(self, "CacheSecurityGroup",
            vpc=vpc,
            description="Security group for ElastiCache",
            allow_all_outbound=True,
        )

        cache_security_group.add_ingress_rule(
            web_security_group,
            ec2.Port.tcp(6379),
            "Allow Redis traffic from web servers"
        )

        cache_cluster = elasticache.CfnCacheCluster(self, "CacheCluster",
            cache_node_type="cache.t3.micro",
            engine="redis",
            num_cache_nodes=1,
            cache_subnet_group_name=cache_subnet_group.ref,
            vpc_security_group_ids=[cache_security_group.security_group_id],
            tags=[{
                "key": "Name",
                "value": f"{environment}-ElastiCache-Cluster",
            }]
        )

        # CloudWatch Alarms
        cloudwatch.Alarm(self, "CPUAlarm",
            alarm_name=f"{environment}-cpu-alarm",
            alarm_description="Alarm for high CPU utilization",
            metric=cloudwatch.Metric(
                namespace="AWS/EC2",
                metric_name="CPUUtilization",
                dimensions_map={
                    "AutoScalingGroupName": auto_scaling_group.auto_scaling_group_name,
                },
                statistic="Average",
            ),
            threshold=80,
            evaluation_periods=2,
            comparison_operator=cloudwatch.ComparisonOperator.GREATER_THAN_THRESHOLD,
        )

        cloudwatch.Alarm(self, "DatabaseCPUAlarm",
            alarm_name=f"{environment}-db-cpu-alarm",
            alarm_description="Alarm for high database CPU utilization",
            metric=cloudwatch.Metric(
                namespace="AWS/RDS",
                metric_name="CPUUtilization",
                dimensions_map={
                    "DBInstanceIdentifier": rds_instance.instance_identifier,
                },
                statistic="Average",
            ),
            threshold=80,
            evaluation_periods=2,
            comparison_operator=cloudwatch.ComparisonOperator.GREATER_THAN_THRESHOLD,
        )

        # Scaling Policy
        auto_scaling_group.scale_on_cpu_utilization("CpuScalingPolicy",
            target_utilization_percent=70,
        )

        # Outputs
        CfnOutput(self, "VpcId",
            description="VPC ID",
            value=vpc.vpc_id,
            export_name=f"{environment}-VpcId",
        )

        CfnOutput(self, "LoadBalancerDNS",
            description="Load Balancer DNS Name",
            value=alb.load_balancer_dns_name,
            export_name=f"{environment}-LoadBalancerDNS",
        )

        CfnOutput(self, "DatabaseEndpoint",
            description="Database Endpoint",
            value=rds_instance.db_instance_endpoint_address,
            export_name=f"{environment}-DatabaseEndpoint",
        )

        CfnOutput(self, "CacheEndpoint",
            description="Cache Cluster Endpoint",
            value=cache_cluster.attr_redis_endpoint_address,
            export_name=f"{environment}-CacheEndpoint",
        )
